README
A Javascript implementation of Steven J. Fortune's algorithm to
efficiently compute Voronoi diagrams. The Voronoi object's purpose is
to solely compute a Voronoi diagram, it is completely standalone, with
no dependency on external code: it contains no rendering code, that is
left to the user of the library. MIT license.

Files
-----

rhill-voronoi-core.js
  Where the Voronoi object is implemented. This is a standalone library,
  there is no dependency.

rhill-voronoi-core.min.js
  The minimized version (using YUI compressor)

rhill-voronoi-demo1.php
rhill-voronoi-demo2.php
rhill-voronoi-demo3.php
  Demo pages to demonstrate usage of the Voronoi object.

excanvas/*
  Used by demo pages
  ExplorerCanvas, giving pre-HTML5 Internet Explorer the ability
  to make sense of HTML5's canvas element. Pulled from
  http://code.google.com/p/explorercanvas/

mootools/*
  Used by rhill-voronoi-demo3.php

Above pages available at
  http://www.raymondhill.net/voronoi/


Main object: Voronoi
--------------------

A Javascript object which allows to compute a Voronoi diagram.
The Voronoi object doesn't render the resulting Voronoi diagram,
the user is responsible for rendering the diagram.


Usage
-----

Roughly:

  var voronoi = new Voronoi();
  var bbox = {xl:0,xr:800,yt:0,yb:600};
  var vertices = [{x:200, y:200}, {x:50, y:250}, {x:400, y:100} /* , ... */];
  // a 'vertex' is an object exhibiting 'x' and 'y' properties. The
  // Voronoi object will add a unique 'voronoiId' property to all
  // vertices. The 'voronoiId' can be used as a key to lookup the
  // associated cell in 'diagram.cells'.
  var diagram = voronoi.compute(vertices, bbox);

The returned 'diagram' variable is a Javascript object with the
following properties:

  diagram.edges:
    An array of unordered, unique Voronoi.Edge objects making up the
    Voronoi diagram. Edges are defined by two vertices, 'va' and 'vb',
	which vertices are shared by connected edges. This mean that
	if you change one vertex belonging to an edge, other connected edges
	will also be changed.

  diagram.cells:
    An array of Voronoi.Cell objects making up the Voronoi diagram. A Cell object
    might have an empty array of halfedges, meaning no Voronoi cell could be computed for a
    particular cell.

  diagram.execTime:
    The time it took to compute the Voronoi diagram, in milliseconds.

Objects:

  Voronoi:
    The Voronoi object which computes a Voronoi diagram.

  Voronoi.Edge:
    lSite: the Voronoi site object at the left of this Voronoi.Edge object.
    rSite: the Voronoi site object at the right of this Voronoi.Edge object (can be null,
      when this is a border edge).
    va: an object with an 'x' and a 'y' property defining the start point
      (relative to the Voronoi site on the left) of this Voronoi.Edge object.
    vb: an object with an 'x' and a 'y' property defining the end point
      (relative to Voronoi site on the left) of this Voronoi.Edge object.

  Voronoi.Cell:
    site: the Voronoi site object associated with the Voronoi cell.
    halfedges: an array of Voronoi.Halfedge objects, ordered counterclockwise, defining the
      polygon for this Voronoi cell.

  Voronoi.Halfedge:
    site: the Voronoi site object owning this Voronoi.Halfedge object.
    edge: a reference to the unique Voronoi.Edge object underlying this Voronoi.Halfedge object.
    getStartpoint(): a method returning an object with an 'x' and a 'y' property for the start
      point of this halfedge. Keep in mind halfedges are always countercockwise.
    getEndpoint(): a method returning an object with an 'x' and a 'y' property for the end
      point of this halfedge. Keep in mind halfedges are always countercockwise.


License
-------

Copyright (c) 2010,2011 Raymond Hill
https://github.com/gorhill/Javascript-Voronoi

Licensed under The MIT License
http://en.wikipedia.org/wiki/MIT_License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
# CHANGELOG.md

## 0.99 (22 Apr 2013):

Now returning all the unique vertices linking all the edges. This addresses
issue #4: https://github.com/gorhill/Javascript-Voronoi/issues/4

## 0.98 (25 Jan 2013):

Added Cell.getBbox() and Cell.pointIntersection() for convenience when using
an external treemap.

## 0.97 (21 Jan 2013):

Merged contribution by Jesse Morgan (https://github.com/morgajel):
Cell.getNeighbourIds()
https://github.com/gorhill/Javascript-Voronoi/commit/4c50f691a301cd6a286359fefba1fab30c8e3b89

## 0.96 (26 May 2011):

Returned diagram.cells is now an array, whereas the index of a cell
matches the index of its associated site in the array of sites passed
to Voronoi.compute(). This allowed some gain in performance. The
'voronoiId' member is still used internally by the Voronoi object.
The Voronoi.Cells object is no longer necessary and has been removed.

## 0.95 (19 May 2011):

No longer using Javascript array to keep track of the beach sections of
the beachline, now using Red-Black tree.

The move to a binary tree was unavoidable, as I ran into finite precision
arithmetic problems when I started to use sites with fractional values.

The problem arose when the code had to find the arc associated with a
triggered Fortune circle event: the collapsing arc was not always properly
found due to finite precision arithmetic-related errors. Using a tree structure
eliminate the need to look-up a beachsection in the array structure
(findDeletionPoint()), and allowed to bring back epsilon down to 1e-9.

## 0.91(21 September 2010):

Lower epsilon from 1e-5 to 1e-4, to fix problem reported at
http://www.raymondhill.net/blog/?p=9#comment-1414

## 0.90 (21 September 2010):

First version.



rhill-voronoi-demo4.php
<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<title>Javascript implementation of Steven Fortune's algorithm to compute Voronoi diagrams: Demo 4</title>
<meta name="Keywords" lang="en" content="voronoi, fortune, javascript, raymond hill"/>
<!--[if lte IE 8]><script type="text/javascript" src="excanvas/excanvas.compiled.js"></script><![endif]-->
<script type="text/javascript" src="rhill-voronoi-core.min.js"></script>
<script type="text/javascript" src="QuadTree.js"></script>
<style type="text/css">
body {font-family:tahoma,verdana,arial;font-size:13px;margin:0;padding:0}
body > div {margin-left:4px;margin-right:4px;}
body > div > div {margin:0;border:1px solid #ccc;border-top:0;padding:4px;}
h1 {margin:0 0 0.5em 0;padding: 4px 5em 4px 4px;font:bold large sans-serif;background-color:#c9d7f1;}
h4 {font-size:14px;margin:0.5em 0 0 0;border:0;border-bottom:solid 1px #c9d7f1;padding:2px;background-color:#e5ecf9;}
#canvasParent {margin-top:0;margin-bottom:1em;padding:0;border:0}
#voronoiCode {font:11px monospace;overflow:auto;color:#666;}
#voronoiCode span {color:green;font-weight:bold;}
</style>
<script id="script" type="text/javascript">
var VoronoiDemo = {
	voronoi: new Voronoi(),
	sites: [],
	diagram: null,
	margin: 0.1,
	canvas: null,
	bbox: {xl:0,xr:800,yt:0,yb:600},
	lastCell: undefined,
	treemap: null,

	init: function() {
		this.canvas = document.getElementById('voronoiCanvas');
		this.randomSites(100,true);
		this.render();
		},

	clearSites: function() {
		this.sites = [];
		this.treemap = null;
		this.diagram = this.voronoi.compute(this.sites, this.bbox);
		this.updateStats();
		},

	randomSites: function(n,clear) {
		if (clear) {this.sites = [];}
		// create vertices
		var xmargin = this.canvas.width*this.margin,
			ymargin = this.canvas.height*this.margin,
			xo = xmargin,
			dx = this.canvas.width-xmargin*2,
			yo = ymargin,
			dy = this.canvas.height-ymargin*2;
		for (var i=0; i<n; i++) {
			this.sites.push({x:self.Math.round((xo+self.Math.random()*dx)*10)/10,y:self.Math.round((yo+self.Math.random()*dy)*10)/10});
			}
		this.treemap = null;
		this.diagram = this.voronoi.compute(this.sites, this.bbox);
		this.updateStats();
		},

	recompute: function() {
		this.treemap = null;
		this.diagram = this.voronoi.compute(this.sites, this.bbox);
		this.updateStats();
		},

	updateStats: function() {
		if (!this.diagram) {return;}
		var e = document.getElementById('voronoiStats');
		if (!e) {return;}
		e.innerHTML = '('+this.diagram.cells.length+' Voronoi cells computed from '+this.diagram.cells.length+' Voronoi sites in '+this.diagram.execTime+' ms &ndash; rendering <i>not</i> included)';
		},

	buildTreemap: function() {
		var treemap = new QuadTree({
			x: this.bbox.xl,
			y: this.bbox.yt,
			width: this.bbox.xr-this.bbox.xl,
			height: this.bbox.yb-this.bbox.yt
			});
		var cells = this.diagram.cells,
			iCell = cells.length;
		while (iCell--) {
			bbox = cells[iCell].getBbox();
			bbox.cellid = iCell;
			treemap.insert(bbox);
			}
		return treemap;
		},

	cellUnderMouse: function(ev) {
		if (!this.diagram) {return;}
		var canvas = document.getElementById('voronoiCanvas');
		if (!canvas) {
			return;
			}
		// >>> http://www.quirksmode.org/js/events_properties.html#position
		var x = 0,
			y = 0;
		if (!ev) {
			ev = window.event;
			}
		if (ev.pageX || ev.pageY) {
			x = ev.pageX;
			y = ev.pageY;
			}
		else if (e.clientX || e.clientY) {
			x = ev.clientX + document.body.scrollLeft + document.documentElement.scrollLeft;
			y = ev.clientY + document.body.scrollTop + document.documentElement.scrollTop;
			}
		// <<< http://www.quirksmode.org/js/events_properties.html#position
		x -= canvas.offsetLeft;
		y -= canvas.offsetTop;
		cellid = this.cellIdFromPoint(x,y);
		if (this.lastCell !== cellid) {
			if (this.lastCell !== undefined) {
				this.renderCell(this.lastCell, '#fff', '#000');
				}
			if (cellid !== undefined) {
				this.renderCell(cellid, '#f00', '#00f');
				}
			this.lastCell = cellid;
			}
		document.getElementById('voronoiCellid').innerHTML = "(" + x + "," + y + ") = " + cellid;
		},

	cellIdFromPoint: function(x, y) {
		// We build the treemap on-demand
		if (this.treemap === null) {
			this.treemap = this.buildTreemap();
			}
		// Get the Voronoi cells from the tree map given x,y
		var items = this.treemap.retrieve({x:x,y:y}),
			iItem = items.length,
			cells = this.diagram.cells,
			cell, cellid;
		while (iItem--) {
			cellid = items[iItem].cellid;
			cell = cells[cellid];
			if (cell.pointIntersection(x,y) > 0) {
				return cellid;
				}
			}
		return undefined;
		},

	renderCell: function(id, fillStyle, strokeStyle) {
		if (id === undefined) {return;}
		if (!this.diagram) {return;}
		var cell = this.diagram.cells[id];
		if (!cell) {return;}
		var ctx = this.canvas.getContext('2d');
		ctx.globalAlpha = 1;
		// edges
		ctx.beginPath();
		var halfedges = cell.halfedges,
			nHalfedges = halfedges.length,
			v = halfedges[0].getStartpoint();
		ctx.moveTo(v.x,v.y);
		for (var iHalfedge=0; iHalfedge<nHalfedges; iHalfedge++) {
			v = halfedges[iHalfedge].getEndpoint();
			ctx.lineTo(v.x,v.y);
			}
		ctx.fillStyle = fillStyle;
		ctx.strokeStyle = strokeStyle;
		ctx.fill();
		ctx.stroke();
		// site
		v = cell.site;
		ctx.fillStyle = '#44f';
		ctx.beginPath();
		ctx.rect(v.x-2/3,v.y-2/3,2,2);
		ctx.fill();
		},

	render: function() {
		var ctx = this.canvas.getContext('2d');
		// background
		ctx.globalAlpha = 1;
		ctx.beginPath();
		ctx.rect(0,0,this.canvas.width,this.canvas.height);
		ctx.fillStyle = 'white';
		ctx.fill();
		ctx.strokeStyle = '#888';
		ctx.stroke();
		// voronoi
		if (!this.diagram) {return;}
		// edges
		ctx.beginPath();
		ctx.strokeStyle = '#000';
		var edges = this.diagram.edges,
			iEdge = edges.length,
			edge, v;
		while (iEdge--) {
			edge = edges[iEdge];
			v = edge.va;
			ctx.moveTo(v.x,v.y);
			v = edge.vb;
			ctx.lineTo(v.x,v.y);
			}
		ctx.stroke();
		// sites
		ctx.beginPath();
		ctx.fillStyle = '#44f';
		var sites = this.sites,
			iSite = sites.length;
		while (iSite--) {
			v = sites[iSite];
			ctx.rect(v.x-2/3,v.y-2/3,2,2);
			}
		ctx.fill();
		},
	};
</script>
</head>
<body onload="VoronoiDemo.init();">
<a href="https://github.com/gorhill/Javascript-Voronoi"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_red_aa0000.png" alt="Fork me on GitHub"></a>
<h1>Javascript implementation of Steven Fortune's algorithm to compute Voronoi diagrams<br/>Demo 4: Looking up a Voronoi cell using a quadtree</h1>
<div id="divroot" style="width:800px;">
<p style="margin-top:0;margin-bottom:0"><a href="/voronoi/rhill-voronoi.php">&lt; Back to main page</a><ul style="margin-top:0">
<li><a href="rhill-voronoi-demo1.php">Demo 1: measuring peformance</a>
<li><a href="rhill-voronoi-demo2.php">Demo 2: a bit of interactivity</a>
<li><a href="rhill-voronoi-demo3.php">Demo 3: Fancy tiling</a>
<li><b>Demo 4: Looking up a Voronoi cell using a quadtree</b>
<li><a href="rhill-voronoi-demo5.html">Demo 5: Lloyd's relaxation</a>
<li><a href="http://www.raymondhill.net/blog/?p=458#comments">Comments</a>
</ul></p>
<h4 class="divhdr">Sites generator</h4>
<div class="divinfo" id="voronoiGenerator">
<input type="button" value="Generate" onclick="VoronoiDemo.randomSites(parseInt(document.getElementById('voronoiNumberSites').value,10),true);VoronoiDemo.render();"/> or <input type="button" value="Add" onclick="VoronoiDemo.randomSites(parseInt(document.getElementById('voronoiNumberSites').value,10),false);VoronoiDemo.render();"/><input id="voronoiNumberSites" type="text" value="100" size="5" maxlength="5"/> sites randomly (Warning: performance might suffer the more sites you add.)
<br/><input id="voronoiClearSites" type="button" value="Clear all sites" onclick="VoronoiDemo.clearSites();VoronoiDemo.render();"/>
</div>
<h4 class="divhdr">Canvas <span id="voronoiStats" style="font:normal 11px sans"></span> <span style="font:normal 11px sans"> / Cell id at <span id="voronoiCellid" style="font:normal 11px sans"></span></span></h4>
<div id="canvasParent">
<noscript>You need to enable Javascript in your browser for this page to display properly.</noscript>
<canvas id="voronoiCanvas" width="800" height="600" onclick="VoronoiDemo.recompute();" onmousemove="VoronoiDemo.cellUnderMouse(event);"></canvas>
<div id="voronoiNoCanvasAlert" style="display:none;padding:1em;background-color:#fcc;color:black;">
<p>Your browser doesn't support the HTML5 &lt;canvas&gt; element technology.</p>
<p>See <a target="_blank" href="http://en.wikipedia.org/wiki/Canvas_(HTML_element)">Wikipedia</a> for information on which browsers support the <u>HTML5 &lt;canvas&gt;</u> technology.</p>
</div>
</div>
<h4 class="divhdr">Javascript source code for this page</h4>
<div class="divinfo" id="voronoiCode">
<pre>
<span>&lt;script type=&quot;text/javascript&quot; src=&quot;<a href="rhill-voronoi-core.js" target="_blank">rhill-voronoi-core.js</a>&quot;&gt;&lt;/script&gt;</span>
<span>&lt;script type=&quot;text/javascript&quot; src=&quot;<a href="https://github.com/mikechambers/ExamplesByMesh/blob/master/JavaScript/QuadTree/src/QuadTree.js" target="_blank">QuadTree.js</a>&quot;&gt;&lt;/script&gt;</span>
...
<div id="scriptContainer"></div>
...
</pre>
</div>
</div>
<script>
(function(){
var srcElem = document.getElementById("script");
if (srcElem) {
    var dstElem = document.getElementById("scriptContainer");
    if (dstElem) {
        dstElem.innerText = srcElem.innerHTML;
        }
    }
})();
</script>
</body>
</html>
